### ISSUE TITLE
Completion of pointers to functions and function objects

### ISSUE THREAD

**Issue by kasos5000** [clangd/clangd#968]
When I'm typing an address of a function, completion appends brackets with signature to the name of the function, although it should complete only the name in this case, according to the syntax.
![completion](https://user-images.githubusercontent.com/79150779/147568047-4c783fcf-e4e9-4f70-8327-203066caf704.PNG)
The same is with function objects:
![func_obj](https://user-images.githubusercontent.com/79150779/147569309-58b7575a-036d-4d90-8653-aee239151067.PNG)
It would be nice if completion inserted brackets with parameters only when needed (when a function is being called) and wrote just a function name when needed (when it is an object). In Qt Creator this feature is already implemented somehow.

**Comment by sam-mccall** [comment]
Hmm, we can't tell if you want to call the function it not: e.g. `&foo(n)` is correct if foo returns a reference and you want a pointer. Even in C: `int *p = &foo(n)->x`.

I suspect which is more common will depend mostly on the codebase. (e.g. I'm fairly sure calling the function is more common in this context in LLVM).

Fine-grained case-by-case heuristics are going to create annoying inconsistencies that disrupt muscle memory.

So I don't see a great way to improve this, any thoughts?

**Comment by HighCommander4** [comment]
In Eclipse CDT, we implemented one heuristic and discussed but did not get around to implementing a second:

  1. If the function names a non-static member of a class, and we are not inside that class's scope, assume the reference will not be a call ([issue](https://bugs.eclipse.org/bugs/show_bug.cgi?id=395562)).
  2. If the function name is being completed in the context of an argument to a function call where the corresponding parameter type is a matching function pointer type, assume the reference will not be a call ([issue](https://bugs.eclipse.org/bugs/show_bug.cgi?id=496563)). We did realize that the reasoning here would need to be heuristic, since the type of the function whose argument is being typed cannot be known accurately until we have a complete argument expression (and thus an argument expression type, and can perform overload resolution).

**Comment by HighCommander4** [comment]
See also #439

**Comment by kasos5000** [comment]
I mentioned that this is implemented in Qt Creator, but it is not exactly so. Qt Creator seems to treat a function object always as an object, not like a call, which is, of course, not a good approach. Also it always treats a function like a pointer to a function when its name begins in an address operator &, which is not always correct as well, because the user might have wanted to enter the address of the returned value (to call the function). So, this is indeed not a trivial problem.

**Comment by tom-anders** [comment]
The use case with Qt is usually something like `QObject::connect(a, &A::foo, b, &B::bar)`, so maybe a heuristic that would be easy to implement and almost always correct would be to just check if we're currently inside a `QObject::connect` call.

But I don't know if creating "dependencies" to 3rd party libs is something we want in clangd?

Implementing the 1st heuristic mentioned by @HighCommander4 would make this one redundant though.

**Comment by HighCommander4** [comment]
> But I don't know if creating "dependencies" to 3rd party libs is something we want in clangd?

We have some precedent in clangd of [special-casing for the Protobuf library](https://searchfox.org/llvm/rev/ca04b49597ccf137da9dfcbb1844d2e8b6a83c07/clang-tools-extra/clangd/SourceCode.cpp#1170-1180).

But since we have an idea of a general heuristic that would handle this case, implementing that is likely preferable.

**Comment by tom-anders** [comment]
> But since we have an idea of a general heuristic that would handle this case, implementing that is likely preferable.

Right, I noticed that we actually would need *both* heuristics to cover the `QObject::connect(a, &A::foo, b, &B::bar)` case though, since it's often called *inside* either A's or B's scope.

**Comment by tom-anders** [comment]
Fix: https://reviews.llvm.org/D134137

**Comment by tom-anders** [comment]
I'm currently taking a look at how to implement the first heuristic mentioned above ("If the function names a non-static member of a class, and we are not inside that class's scope, assume the reference will not be a call"). I'm a bit stuck at how to figure out the current class scope (from inside `CodeCompletionBuilder`) - Looks like maybe I could use Sema's `CurContext*` field for this and use `getParent()` until we find a class (or reach nullptr)? Am I  on the right track here @HighCommander4?

**Comment by HighCommander4** [comment]
> Looks like maybe I could use Sema's `CurContext*` field for this and use `getParent()` until we find a class (or reach nullptr)? Am I on the right track here @HighCommander4?

I don't have super familiar with this either, but yes, that sounds right to me.

**Comment by tom-anders** [comment]
Here's another heuristic that's probably a bit easier to implement than the 2nd one mentioned above:

For `&Foo::^`, if the completed function returns `void`, it can't be a call (this is even true regardless of the current scope). This would be ideal for Qt, where signals (always) and slots (usually) return `void`.

**Comment by HighCommander4** [comment]
> For `&Foo::^`, if the completed function returns `void`, it can't be a call

Good point. I believe we can even extend this to built-in types returned by value or pointer (the `&` would necessary pertain to the return object itself and not a field of it, and you'd be taking the address of a temporary).

**Comment by tom-anders** [comment]
Sounds good! Right now I'm struggling a bit with detecting inside Sema whether there actually is a `&` or not, I couldn't find anything related yet neither in `ResultBuilder` nor in `CodeCompletionResult`

**Comment by tom-anders** [comment]
Ah, `clangd::Parser::ParseCastExpression` (which according to my debugger appears in the stacktrace when completing `&Foo::^`) seems to have a `isAddressOfOperand` parameter that looks like it does what we want.
